R/utils-cut.R

```diff
@@ -0,0 +1,225 @@
+      if (anyDuplicated(breaks) > 0) {
+        next
+      }
+
+      # calculate the inter-break differences (binWidths) from rounded values
+      diffs <- diff(breaks)
+
+      # if only one difference (e.g. one bin), we can't do the following part
+      # which checks that the differences are not too different from each other
+      if (length(diffs) < 2) {
+        break
+      }
+
+      # now see if the normalised variance (index of dispersion) of the diffs is small enough to
+      # stop increasing the number of significant digits
+      dispersion <- var(diffs)/mean(diffs)
+      if (dispersion < 1e-05) {
+        break
+      }
+    }
+  }
+
+  # Now, we know we have enough bins to encompass min_x and max_x, but they could have been
+  # rounded to slightly just above or below, respectively.
+  # If so, we will adjust the first and last bins by subtracting/adding one "least significant digit"
+  # from them.
+
+  if (min_x < breaks[1]) {
+    newStart <- breaks[1] - signifDigitEpsilon(breaks[1], requiredDigits)
+    breaks[1] <- as.numeric(formatC(newStart, digits = requiredDigits, width = 1L))
+  }
+
+  lastBreak = breaks[length(breaks)]
+  if (max_x > lastBreak) {
+    newEnd <- lastBreak + signifDigitEpsilon(lastBreak, requiredDigits)
+    breaks[length(breaks)] <- as.numeric(formatC(newEnd, digits = requiredDigits, width = 1L))
+  }
+
+  # safety checks
+  if (min_x < breaks[1]) {
+    stop("Fatal problem with cut_width start bin extension")
+  }
+  if (max_x > breaks[length(breaks)]) {
+    stop("Fatal	problem	with cut_width end bin extension")
+  }
+
+  cut(x, breaks, include.lowest = TRUE, right = (closed == "right"), dig.lab = requiredDigits, ...)
+}
+
+# Find the left side of left-most bin
+find_origin <- function(x_range, width, boundary) {
+  shift <- floor((x_range[1] - boundary) / width)
+  boundary + shift * width
+}
+
+breaks <- function(x, method = c('equalInterval', 'quantile', 'sd'), nbins = NULL, binwidth = NULL) {
+  method <- veupathUtils::matchArg(method)
+  if ((!is.null(nbins) && !is.null(binwidth)) || (is.null(nbins) && is.null(binwidth))) {
+    stop("Specify exactly one of n and width")
+  }
+
+  rng <- range(x, na.rm = TRUE, finite = TRUE)
+
+  if (method == "equalInterval") {
+    if (!is.null(binwidth)) {
+      brks <- scales::fullseq(rng, binwidth)
+    } else {
+      brks <- seq(rng[1], rng[2], length.out = nbins + 1)
+    }
+  } else if (method == "quantile") {
+    if (!is.null(binwidth)) {
+      probs <- seq(0, 1, by = binwidth)
+    } else {
+      probs <- seq(0, 1, length.out = nbins + 1)
+    }
+    brks <- stats::quantile(x, probs, na.rm = TRUE)
+  } else {
+    warning("`breaks` method `sd` does NOT currently support specifying `nbins` or `binwidth`.")
+    med <- median(x)
+    sd <- sd(x)
+    brks <- c(min(x), med-(sd*2), med-sd, med, med+sd, med+(sd*2), max(x))
+ 
+    valid <- all(unlist(lapply(2:length(brks), FUN = function(x) {brks[[x]] > brks[[x-1]]})))
```

Comment:
How is this different from the `unique(brks)` below?

I'm sure it is. I just don't see how!

Reply:
I actually think you're right that the unique isn't necessary. I wrote that validation first and realized it wasn't sufficient but then never removed it after adding the other